File: coauthor/agent/MergeAgent.py
import os
import re

from ..logger import logger

from .agent_config import AgentConfig
from .agent_dataclass import AgentSetting, AgentPrompt
from .DirectAgent import DirectAgent
from .agent_state import AgentStateRound, AgentStateGlobal
from .model_handler import ModelHandler


class MergeAgent(DirectAgent):
    """Agent for merging multiple edited files into a single output."""

    def __init__(
        self,
        modelHandler: ModelHandler,
        agentConfig: AgentConfig,
        agentSetting: AgentSetting,
        agentPrompt: AgentPrompt,
        agentPath: str,
    ) -> None:
        """Initialize merge agent with model handler, configs, settings, prompts and path."""
        super().__init__(modelHandler, agentConfig, agentSetting, agentPrompt, agentPath)
        self.outputFile = [self.getOutputFile(r) for r in range(2)]

    def _parseFilenameParts(self, editedBase: str) -> tuple[str, str, int, str]:
        """Parse filename parts to extract base name, agent, round number and model."""
        parts = editedBase.split("_")
        underscoreCount = editedBase.count("_")
        base = parts[0]

        # Extract agent name
        agent = self._extractAgentName(parts, underscoreCount)
        if agent is None:
            raise ValueError(f"Could not extract agent name from edited base: {editedBase}")

        # Extract round number
        roundMatch = re.search(r"_r(\d+)_", editedBase)
        if not roundMatch:
            raise ValueError(f"Could not extract round number from edited base: {editedBase}")
        roundNum = int(roundMatch.group(1))

        # Get model name (last part)
        model = parts[-1]

        return base, agent, roundNum, model

    def getOutputFile(self, currRound: int) -> str:
        """Generate output filename for merged content."""
        inputFile = self.agentConfig.inputFile
        editedFile = self.agentConfig.editedFile

        if not editedFile:
            raise ValueError("editedFile must be specified for merge handler")

        inputDir = os.path.dirname(inputFile)
        inputBase, _ = os.path.splitext(os.path.basename(inputFile))
        editedBase, _ = os.path.splitext(os.path.basename(editedFile))

        # Parse filename components
        base, agent, roundNum, model = self._parseFilenameParts(editedBase)

        # Use original input base if it differs from edited base
        if inputBase != base:
            base = inputBase

        # Construct output filename
        outputFile = f"{base}_{agent}_r{roundNum}_full_{model}.tex"
        output_path = os.path.join(inputDir, outputFile)
        logger.info(f"Merge output file: {output_path}")
        return output_path

    def _extractAgentName(self, parts: list[str], underscoreCount: int) -> str | None:
        """Extract agent name from filename parts.

        Handles two formats:
        - Standard: base_agent_r1_model
        - Complex: MutualInfo_restructured_polish_r1_sonnet++
        """
        if underscoreCount == 3:
            # Standard format
            return parts[1]

        # Complex format - collect parts until round number
        agent_parts = []
        for i, part in enumerate(parts[1:], 1):
            if part.startswith("r") and part[1:].isdigit():
                return "_".join(agent_parts)
            agent_parts.append(part)
        return None

    def handleOutput(
        self,
        stateRound: AgentStateRound,
        stateGlobal: AgentStateGlobal,
        outputFile: str,
        endTurn: bool,
        currRound: int = 0,
    ) -> list[str]:
        """Process and handle output files for the current round."""
        if endTurn:
            _files = super().handleOutput(stateRound, stateGlobal, outputFile, endTurn, currRound)
            logger.info(f"Output file: {outputFile}")
            return _files
        return []

File: coauthor-for-vs-code/src/agent/MergeAgent.ts
// Standard library imports
import * as path from 'path';

// Local imports - agent components
import { DirectAgent } from './DirectAgent';
import { AgentStateRound, AgentStateGlobal } from './AgentState';
import { ModelHandler } from './ModelHandler';
import { AgentConfig } from './AgentConfig';
import { AgentSetting, AgentPrompt } from './AgentDataclass';

/**
 * Specialized agent for merging multiple edited files into a consolidated output.
 * Handles complex filename parsing and maintains file relationships during merging.
 */
export class MergeAgent extends DirectAgent {
  constructor(
    modelHandler: ModelHandler,
    agentConfig: AgentConfig,
    agentSetting: AgentSetting,
    agentPrompt: AgentPrompt,
    agentPath: string,
  ) {
    super(modelHandler, agentConfig, agentSetting, agentPrompt, agentPath);
    this.outputFile = [this.getOutputFile(0), this.getOutputFile(1)];
  }

  /**
   * Extracts components from edited filename for merge operations.
   * @param editedBase Base name of edited file without extension
   * @returns Tuple of [base name, agent name, round number, model name]
   * @throws Error if filename components cannot be extracted
   */
  private parseFilenameParts(
    editedBase: string,
  ): [string, string, number, string] {
    const parts = editedBase.split('_');
    const underscoreCount = editedBase.split('_').length - 1;
    const base = parts[0];

    // Extract agent name
    const agent = this.extractAgentName(parts, underscoreCount);
    if (!agent) {
      throw new Error(
        `Could not extract agent name from edited base: ${editedBase}`,
      );
    }

    // Extract round number
    const roundMatch = editedBase.match(/_r(\d+)_/);
    if (!roundMatch) {
      throw new Error(
        `Could not extract round number from edited base: ${editedBase}`,
      );
    }
    const roundNum = parseInt(roundMatch[1], 10);

    // Get model name (last part)
    const model = parts.at(-1) || '';

    return [base, agent, roundNum, model];
  }

  /**
   * Generates output filename for merged content based on input and edited files.
   * @param currRound Current round number (unused in merge operations)
   * @returns Path to output file for merged content
   * @throws Error if editedFile is not specified
   */
  protected getOutputFile(currRound: number): string {
    const inputFile = this.agentConfig.inputFile;
    const editedFile = this.agentConfig.editedFile;

    if (!editedFile) {
      throw new Error('editedFile must be specified for merge handler');
    }

    const inputDir = path.dirname(inputFile);
    const inputBase = path.parse(inputFile).name;
    const editedBase = path.parse(editedFile).name;

    // Parse filename components
    const [base, agent, roundNum, model] = this.parseFilenameParts(editedBase);

    // Use original input base if it differs from edited base
    const finalBase = inputBase !== base ? inputBase : base;

    // Construct output filename
    const outputFile = `${finalBase}_${agent}_r${roundNum}_full_${model}.tex`;
    const outputPath = path.join(inputDir, outputFile);
    return outputPath;
  }

  /**
   * Extracts agent name from filename parts handling multiple formats.
   * @param parts Array of filename parts split by underscore
   * @param underscoreCount Total number of underscores in filename
   * @returns Agent name or null if not found
   */
  private extractAgentName(
    parts: string[],
    underscoreCount: number,
  ): string | null {
    if (underscoreCount === 3) {
      // Standard format
      return parts[1];
    }

    // Complex format - collect parts until round number
    const agentParts: string[] = [];
    for (let i = 1; i < parts.length; i++) {
      const part = parts[i];
      if (part.startsWith('r') && /^\d+$/.test(part.slice(1))) {
        return agentParts.join('_');
      }
      agentParts.push(part);
    }
    return null;
  }

  /**
   * Processes output files for merge operation.
   * @param stateRound Current round state
   * @param stateGlobal Global conversation state
   * @param outputFile Path to the output file
   * @param endTurn Whether this is the end of the current turn
   * @param currRound Current round number (defaults to 0)
   * @returns Array of processed output file paths
   */
  protected async handleOutput(
    stateRound: AgentStateRound,
    stateGlobal: AgentStateGlobal,
    outputFile: string,
    endTurn: boolean,
    currRound: number = 0,
  ): Promise<string[]> {
    if (endTurn) {
      this.logger.debug(`Processing merge output for round ${currRound}`);
      const files = await super.handleOutput(
        stateRound,
        stateGlobal,
        outputFile,
        endTurn,
        currRound,
      );
      this.logger.info(`Merge output file: ${outputFile}`);
      return files;
    }
    return [];
  }
}
